Create initial API structure

actuallyatiger opened the pull request:
Closes #5

I've created an initial representation of the API. Data is cached locally inside dataframes with several filter and getter methods.

I've taken a conservative approach to removing columns as this can easily be changed and gives the most flexibility to algorithm designers.

Several parts of the data have nested responses. In these cases I've separated the nested fields into their own dataframe with a link back to the origin's ID (e.g. booking <- transport <- route)

Additionally, I have uploaded example data to the shared drive so you can see the structure it takes.

Please let me know what other filters you would like that I may have missed, or any other suggestions which you think may help you.

actuallyatiger commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,609 @@
+            routes.append(self.getRoutesForTransport(transport.id))
+        return pd.concat(routes) if routes else pd.DataFrame()
+
+    def getBookingsByCargoWindow(self, start: str, end: str):
```
Currently dates/times are stored as strings which is what the API returns. This means it's easier in some cases to search/filter but on the other hand it could be more useful to use the actual pandas datetime type.

It would be fairly simple for me to convert to datetime objects if you believe this would be useful.

micnekr replied:
I think being able to compare 2 datetimes would be really useful in the algorithms, so it would be great to see that.

micnekr commented:
Hi, @actuallyatiger, here is an issue I am getting:
When importing src.api or src.api.SquidAPI, there is a circular import:
```
ImportError                               Traceback (most recent call last)
Cell In[2], line 1
----> 1 from src.api import *

File ~/Documents/code/python/chameleon/src/api/__init__.py:1
----> 1 from .SquidAPI import SquidAPI
      2 import os
      3 from dotenv import load_dotenv

File ~/Documents/code/python/chameleon/src/api/SquidAPI.py:4
      2 import pandas as pd
      3 import os
----> 4 from . import API_KEY
      7 class SquidAPI:
      8     """
      9     API client for interacting with Squid Software's logistics API.
     10     Handles data fetching, caching, and provides access methods for logistics entities.
     11     """

ImportError: cannot import name 'API_KEY' from partially initialized module 'src.api' (most likely due to a circular import) (/home/<user>/Documents/code/python/chameleon/src/api/__init__.py)

In [3]: from src.api.SquidAPI import *
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In[3], line 1
----> 1 from src.api.SquidAPI import *

File ~/Documents/code/python/chameleon/src/api/__init__.py:1
----> 1 from .SquidAPI import SquidAPI
      2 import os
      3 from dotenv import load_dotenv

File ~/Documents/code/python/chameleon/src/api/SquidAPI.py:4
      2 import pandas as pd
      3 import os
----> 4 from . import API_KEY
      7 class SquidAPI:
      8     """
      9     API client for interacting with Squid Software's logistics API.
     10     Handles data fetching, caching, and provides access methods for logistics entities.
     11     """

ImportError: cannot import name 'API_KEY' from partially initialized module 'src.api' (most likely due to a circular import) (/home/<user>/Documents/code/python/chameleon_rust/src/api/__init__.py)

```

micnekr commented:
You have a few arguments treated as optional, but that is not reflected in the types (for example, `day` in `__fetchBookings`. Can you please use e.g. `Optional[str]` in this case?

actuallyatiger commented:
Converted obvious fields to datetime. Need opinions on dates and individual times (maybe use `datetime.date` and `datetime.timedelta`)

Also added OSRM methods to close #7, let me know if there's anything else you'd like.

micnekr commented:
Hi, sorry, I didn't realise that you added me as a reviewer. I will have a look at it now

micnekr commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+    def getBookingsByPickupWindow(self, start: datetime, end: datetime) -> pd.DataFrame:
+        """Find bookings with pickup windows within the specified time range"""
+        return self.bookings[
+            (self.bookings.notna(start) and self.bookings.notna(end))
```
In situations like these here and elsewhere in your code, the result might be [not what you expect](https://stackoverflow.com/questions/21415661/logical-operators-for-boolean-indexing-in-pandas#21415990), and probably not what you meant. I think it makes sense to search for "and" and replace it with `&` where necessary.

micnekr commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+        )
+
+    def getCoordDist(
+        self, coords: List[List[float]]
```
A `List[Tuple[float, float]]` might be more suitable here

actuallyatiger replied:
I would agree - in fact it was originally this. However the OSRM route explicitly wants a "coordinates=[[...],[...]]" syntax and was errored for parenthesis e.g. "[(...),(...)]", so I stuck with a list instead of tuple as otherwise I would need to pre-process and either convert them all to lists or create an extra function to convert them to the correct format instead of letting python do it automatically. (Can do this if you believe passing them as tuples helps though)

micnekr commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+
+    def getCodeDist(
+        self, codes: List[str]
+    ) -> Optional[List[Union[float, List[float]]]]:
```
The return type here should be the same as in `getCoordDist`

micnekr commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+            route_stops.append([route["index"], route.location_id])
+
+        # Sort routes by index
+        route_stops = sorted(route_stops, key=lambda x: x[0])
```
Wait, why are they stored in a potentially unsorted order in the dataframe?

actuallyatiger replied:
The response from the API is unsorted. I would sort it when creating the dataframe however as it's planned that the data we're acting on can change and our plan can change with it, I thought it'd be best to prevent frequently re-sorting the data when it changes.

micnekr commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+        return self.getCoordDist(coords)
+
+    def getCoordMatrix(
+        self, coords: List[List[float]]
```
Same comment about `Tuple[float, float]` applies here

micnekr commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+        self.__OSRM_URL = "https://osrm.squid.software"
+
+        # Define paths for cached data files
+        self.__LOCATIONPATH = "../../data/locations.csv"
```
Python will look at imports relative to the original directory from which the code is started; because of import issues, it is a lot easier to have it be the root directory of the project. So this should probably be `./data/locations.csv`, etc.

Even then, it would be nice to try to create this directory in-code so the code doesn't fail with `OSError: Cannot save file into a non-existent directory: '../../data'`

micnekr commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+                inplace=True,
+            )
+            self.trucks["number"] = self.trucks["number"].astype(int)
+            self.trucks.to_csv(self.__TRUCKPATH)
```
I'm getting `TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'` here

actuallyatiger replied:
Do trucks in the data you're using have a `number` field? All trucks in the API data (should, and I have confirmed this with the data in for our trial) have this field to a unique number. It is actually intended that it raises an error when there is a truck missing a value here.

actuallyatiger replied:
I've been using the frozen data. Interesting that the general data is missing this, so I'll need to look into this.

actuallyatiger commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+        self.__fetchTruckStarts()
+        self.__fetchDrivers()
+        self.__fetchChassis()
+        self.__fetchBookings(day)
```
Seems like i forgot to change day over when converting to datetimes, will fix.

actuallyatiger commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+            self.truck_starts.set_index("id", inplace=True)
+        else:
+            raise ValueError(
+                "Truck starts data not found. Please fetch the data first."
```
This was meant as a more temporary case for now. This method handles data I've received from Christian and isn't accessible from the API as it changes daily. The idea is likely that your methods will track the state of the drivers as they are ran for several days, so this is just there to give an initial starting point for when we're testing this method.

Will add a comment about this.

actuallyatiger commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+    def __fetchBookings(self, day: Optional[date] = None):
+        """
+        Fetch or load booking data and related transport/route information.
+        Optionally filter by date of first pickup.
```
Will only fetch bookings from the API in which the first pickup is on or after this date (the initial idea was to do this and then filter the dataframe to a specific day but that won't work for the complexity of some transports, so this seemed like an acceptable middle ground.

micnekr replied:
I meant that it would be useful to add this explanation to the docstring

actuallyatiger commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+
+        for booking in res:
+            # Process container information
+            if booking.get("container"):
```
The purpose of this is that if a container is assigned it gets the necessary describing info (id and number) and stores it directly instead of handling nested objects. If there is no assigned container, these fields should be `None` (there's a small mistake here so I will push a fix). This section of code is pretty separate from `container_type` otherwise and that is left in the booking data regardless of if there is a container assigned already

actuallyatiger commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+            all_bookings.append(booking)
+
+        # Create dataframes and drop unwanted fields
+        if all_bookings:
```
If there are no bookings (for example you're trying to plan too far ahead so no bookings have been made yet), this catches this case and results in an empty dataframe as the `set_index` etc will otherwise throw an error.

actuallyatiger commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+    # BOOKING GETTERS
+    def getBookings(
+        self,
+        code: str = None,
```
Haha thought I caught them all, will double check, thanks.

actuallyatiger commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,829 @@
+    def getBookingsByCargoWindow(self, start: datetime, end: datetime) -> pd.DataFrame:
+        """Find bookings with cargo windows within the specified time range"""
+        return self.bookings[
+            (self.bookings.notna(start) and self.bookings.notna(end))
```
Appears to be a discrepency between pandas/python versions, as this seems to have been deprecated and removed. Will fix.

micnekr commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,860 @@
+        Optionally filter by date of first pickup.
+        """
+        # Check if the data has already been fetched
+        if (
```
You are only filtering by `day` if the bookings aren't cached. Since it is not obvious when it would and wouldn't filter, doing this will catch someone out at some point. Please filter out for `day` in both cases

micnekr commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,860 @@
+    def getTransports(
+        self,
+        adr: Optional[bool] = None,
+        area: Optional[str] = None,
```
Can you please change the default to ask for a `port` transport? I almost used the default arguments until you reminded us that we are only looking at port transports, and I don't want anyone to make that mistake by accident. What about the default being `"port"` and to opt-out and look for all transports, you can search for `area="all"`?

micnekr commented on src/api/SquidAPI.py:
```diff
@@ -0,0 +1,860 @@
+
+    def getCoordDist(
+        self, coords: List[List[float]]
+    ) -> Optional[Dict[str, Union[float, List[Dict[str, float]]]]]:
```
This is not the shape of the response, see OSRM docs
